feat: Rd 12832 storm otel instrumentation (#395)
lumigo-io/opentelemetry-java-distro: instrumentation/storm/README.md

# Background on storm instrumentation

This is lumigo instrumentation for the [storm](https://storm.apache.org/) library.

> ## We instrument 3 things in the storm library:

Reviewer: can you please specify the OTEL attributes you've used?

# Background on storm instrumentation

This is lumigo instrumentation for the [storm](https://storm.apache.org/) library.

## We instrument 3 things in the storm library:

* Storm Spouts 
* Storm Bolts 
* Storm `ExecutorTransfer` - the class that is responsible for transferring tuples between storm components (used inside the storm `OutputCollector`).

each one of those components has their own instrumentation class.

> I create an internal span for each run of the spout / bolt and for each tuple that is emitted with the OutputCollector to each component its emitted to.

Reviewer: ```suggestion
An internal span is created for each run of the `spout` / `bolt` and each tuple emitted with the OutputCollector to each component it's emitted to.
```

## Otel conventions

> for this instrumentation add following attributes to the spans:

Reviewer: ```suggestion
for this instrumentation we used the following attributes:
```

1. `thread.name` - semantic convention default attribute
2. `messaging.system` - semantic convention default attribute
3. `messaging.message.id` - semantic convention default attribute
4. `messaging.destination.name` - semantic convention default attribute
5. `storm.type` - the type of the storm component spout / bolt
6. `storm.id` - the id of the storm topology
7. `storm.version` - the version of the storm library
8. `storm.tuple.values` - the value that emitted to each bolt
9. `source.component` - the previous component name (can be used to create trigger by behaviour)
10. `component.name` - the name of the component in the invocation override the otel_service_name property

## Building transactions

> we can build the transaction because we add the storm message id to the spout / bolt span attributes as `messaging.message.id` (based on Opentelemetry semantic convensions).

Reviewer: ```suggestion
Transactions are built using storm `message.id` added to the `spout` / `bolt` span attributes (based on Opentelemetry semantic convensions).
```

> this message id is created by the `ExecutorTransfer` when it sends the tuple to the next component, and we add it to the `ExecutorTransfer` span too.

Reviewer: ```suggestion
This message id is created by the `ExecutorTransfer` when it sends the tuple to the next component, and we add it to the `ExecutorTransfer` span too.
```

In addition, the `ExecutorTransfer` span has a parent span of the spout / bolt that emitted the tuple.

> 

Reviewer: consider adding a mini diagram?

## Possible improvements:

> * We create internal span for spout that doesn't emit any tuple.

Reviewer: avoid using the term "we".

```suggestion
* Create Internal span for spout is which emitted no tuple. 
```
?

* We create internal span for tuples that emitted with the `OutputCollector` to storm internal components (like the acker) - we don't show them in the three because they are not bolts / spout, so they don't have spans.

> * We infer the component name from the thread name because I didn't manage to find any other way to do it.

Reviewer: ```suggestion
* We infer the component name from the thread name because no other way was found
```
